### PR TITLE
Update foreach APIs to use scalar lists

### PR DESCRIPTION
**Pull request by izdeby**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52450 wip
* #49039 [wip] Replace optimizers in torch.optim with the ones from torch.optim._multi_tensor
* #52449 Add type promotion to a fast path for foreach APIs
* #52448 Support broadcasting via slow path
* #48224 [wip] Pushed unsupported scenarios to slow path
* #51170 Updated foreach pointwise ops to support complex and bloat16 on CUDA
* #49714 Update foreach max/min
* #51060 Refactor foreach pointwise ops tests to use OpInfo
* #51059 Refactor foreach binary ops tests with tensor lists to use OpInfo
* #51058 Refactor foreach binary ops tests with scalars to use OpInfo
* #49712 Refactor foreach unary ops tests to use OpInfo
* #49250 Update foreach binary ops with a single scalar and list of tensors
* #49249 Update foreach binary ops with a scalar list
* #51131 Refactor ForeachUtils.h
* #51894 Refactor ForeachUnaryOp.cu
* **#51893 Update foreach APIs to use scalar lists**

Differential Revision: [D26323606](https://our.internmc.facebook.com/intern/diff/D26323606)

------------------------------------------
Motivation
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef

Testing
Update the tests assuming that any scalar type can be passed now, not just double.

### CODE REVIEW

File: test/backward_compatibility/check_backward_compatibility.py [hunk 1]
```diff
@@ -61,18 +61,18 @@
...
-    ("aten::_foreach_addcdiv_", datetime.date(2021, 2, 25)),
-    ("aten::_foreach_add_", datetime.date(2021, 2, 25)),
-    ("aten::_foreach_mul", datetime.date(2021, 2, 25)),
-    ("aten::_foreach_addcmul", datetime.date(2021, 2, 25)),
-    ("aten::_foreach_add", datetime.date(2021, 2, 25)),
-    ("aten::_foreach_sub_", datetime.date(2021, 2, 25)),
-    ("aten::_foreach_div", datetime.date(2021, 2, 25)),
-    ("aten::_foreach_div_", datetime.date(2021, 2, 25)),
-    ("aten::_foreach_addcdiv", datetime.date(2021, 2, 25)),
-    ("aten::mkldnn_linear", datetime.date(2021, 3, 2)),
```
Comment (gchanan) [inline test/backward_compatibility/check_backward_compatibility.py:75]: what is happening with mkldnn_linear here?

File: aten/src/ATen/native/ForeachUtils.h [hunk 2]
```diff
@@ -128,7 +119,7 @@ bool can_use_fast_route(TensorList tensors, Scalar scalar) {
       return false;
     }
 
-    if (!will_promote_tensor(t, scalar)) {
+    if (will_promote_tensor(t, scalar)) {
```
Comment (gchanan) [inline aten/src/ATen/native/ForeachUtils.h:122]: wait, was this computing the wrong thing before?

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -6749,49 +6749,49 @@
     CPU: foreach_tensor_div_list_kernel_slow_
     CUDA: foreach_tensor_div_list_kernel_cuda_
 
-- func: _foreach_add.ScalarList(Tensor[] tensors, float[] scalars) -> Tensor[]
+- func: _foreach_add.ScalarList(Tensor[] tensors, Scalar[] scalars) -> Tensor[]
```
Comment (iseeyuan) [inline aten/src/ATen/native/native_functions.yaml:6752]: I don't think it breaks backward compatibility. Could you confirm? What would happen if a scalar other than float is passed into an old model, in the new code, unlikely?

File: aten/src/ATen/native/ForeachUtils.h (hunk 2 above)
Comment (izdeby) [reply aten/src/ATen/native/ForeachUtils.h:124]: no, the name was wrong

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (izdeby) [reply aten/src/ATen/native/native_functions.yaml:6754]: no BC because those are private unused APIs.

**Comment by izdeby** [comment]
@ngimel, fixed the build! This is the [fix](https://github.com/pytorch/pytorch/pull/51893/commits/c70832e4444afdba3a8183820cc554a062813e1c). The problem was was incorrect check for strides which caused some cases to go via fast path when it should have gone via slow.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**